source/common/stats/thread_local_store.cc

```diff
@@ -208,31 +255,143 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
   // See comments in counter(). There is no super clean way (via templates or otherwise) to
   // share this code so I'm leaving it largely duplicated for now.
   std::string final_name = prefix_ + name;
-  HistogramSharedPtr* tls_ref = nullptr;
+  ParentHistogramSharedPtr* tls_ref = nullptr;
+
   if (!parent_.shutting_down_ && parent_.tls_) {
-    tls_ref =
-        &parent_.tls_->getTyped<TlsCache>().scope_cache_[this->scope_id_].histograms_[final_name];
+    tls_ref = &parent_.tls_->getTyped<TlsCache>()
+                   .scope_cache_[this->scope_id_]
+                   .parent_histograms_[final_name];
   }
 
   if (tls_ref && *tls_ref) {
     return **tls_ref;
   }
 
   std::unique_lock<std::mutex> lock(parent_.lock_);
-  HistogramSharedPtr& central_ref = central_cache_.histograms_[final_name];
+  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[final_name];
   if (!central_ref) {
     std::vector<Tag> tags;
     std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
-    central_ref.reset(
-        new HistogramImpl(final_name, parent_, std::move(tag_extracted_name), std::move(tags)));
+    central_ref.reset(new ParentHistogramImpl(final_name, parent_, *this,
+                                              std::move(tag_extracted_name), std::move(tags)));
   }
 
   if (tls_ref) {
     *tls_ref = central_ref;
   }
-
   return *central_ref;
 }
 
+Histogram& ThreadLocalStoreImpl::ScopeImpl::tlsHistogram(const std::string& name,
+                                                         ParentHistogramImpl& parent) {
+  // See comments in counter() which explains the logic here.
+
+  // Here prefix will not be considered because, by the time ParentHistogram calls this method
+  // during recordValue, the prefix is already attached to the name.
+  TlsHistogramSharedPtr* tls_ref = nullptr;
+  if (!parent_.shutting_down_ && parent_.tls_) {
+    tls_ref = &parent_.tls_->getTyped<TlsCache>().scope_cache_[this->scope_id_].histograms_[name];
+  }
+
+  if (tls_ref && *tls_ref) {
+    return **tls_ref;
+  }
+
+  std::unique_lock<std::mutex> lock(parent_.lock_);
```

Comment:
I don't think we need to take this lock here.  Kinda unrelated side note: I think we can make the `getTagsForName` method on `ThreadLocalStore` `const`.

Reply:
Yup that's right. I think I also noticed that at some point and then forgot about it amidst all the issues. Thanks for mentioning it. I will see if I can nurse my computer into making some of these small changes.

Reply:
@mattklein123 if you have trouble with your computer, merge this PR and I can do the follow-up PR addressing these minor comments if that’s ok.

Reply:
Sure, that's fine. @htuch can you approve/merge if this looks good?